From 7c0e7a156056122ecb40fd1302457758c05f1a73 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:10:55 -0400
Subject: [usb] pl2303: add helper function for direct baud-rate encoding

Message-id: <1410887463-20674-584-git-send-email-dzickus@redhat.com>
Patchwork-id: 95395
O-Subject: [RHEL7 PATCH 583/591] USB: pl2303: add helper function for direct baud-rate encoding
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 20b4c787199f4fcf0fae5ed78a4ff0104e2afaa3
Author: Johan Hovold <jhovold@gmail.com>
Date:   Sun Dec 29 19:23:14 2013 +0100

    USB: pl2303: add helper function for direct baud-rate encoding

    Add helper function for direct baud-rate encoding.

    Signed-off-by: Johan Hovold <jhovold@gmail.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/serial/pl2303.c b/drivers/usb/serial/pl2303.c
index 3db85a9..48e63fd 100644
--- a/drivers/usb/serial/pl2303.c
+++ b/drivers/usb/serial/pl2303.c
@@ -341,6 +341,18 @@ static speed_t pl2303_get_supported_baud_rate(speed_t baud)
  return baud;
 }
 
+/*
+ * NOTE: If unsupported baud rates are set directly, the PL2303 seems to
+ *       use 9600 baud.
+ */
+static speed_t pl2303_encode_baud_rate_direct(unsigned char buf[4],
+        speed_t baud)
+{
+ put_unaligned_le32(baud, buf);
+
+ return baud;
+}
+
 static speed_t pl2303_encode_baud_rate_divisor(unsigned char buf[4],
         speed_t baud)
 {
@@ -378,16 +390,12 @@ static void pl2303_encode_baud_rate(struct tty_struct *tty,
 
  if (spriv->type->max_baud_rate)
   baud = min_t(speed_t, baud, spriv->type->max_baud_rate);
- /*
-  * Set baud rate to nearest supported value.
-  *
-  * NOTE: If unsupported values are set directly, the PL2303 seems to
-  *       use 9600 baud.
-  */
+
+ /* Set baud rate to nearest supported value. */
  baud = pl2303_get_supported_baud_rate(baud);
 
  if (baud <= 115200)
-  put_unaligned_le32(baud, buf);
+  baud = pl2303_encode_baud_rate_direct(buf, baud);
  else
   baud = pl2303_encode_baud_rate_divisor(buf, baud);
 
-- 
1.7.1